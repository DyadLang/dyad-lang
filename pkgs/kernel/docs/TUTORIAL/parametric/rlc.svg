<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 1000 1000"
    overflow="visible" shape-rendering="geometricPrecision" text-rendering="geometricPrecision">
    <defs>
        <filter id="red-shadow" color-interpolation-filters="sRGB">
            <feDropShadow dx="0" dy="0" stdDeviation="100" flood-color="#ff0000" flood-opacity="0.5"></feDropShadow>
        </filter>
        <filter id="green-shadow" color-interpolation-filters="sRGB">
            <feDropShadow dx="0" dy="0" stdDeviation="100" flood-color="#00ff00" flood-opacity="0.5"></feDropShadow>
        </filter>
        <filter id="blue-shadow" color-interpolation-filters="sRGB">
            <feDropShadow dx="0" dy="0" stdDeviation="100" flood-color="#0000ff" flood-opacity="0.5"></feDropShadow>
        </filter>
        <filter id="drop-shadow" color-interpolation-filters="sRGB">
            <feDropShadow dx="0" dy="0" stdDeviation="40" flood-opacity="0.5"></feDropShadow>
        </filter>
    </defs>
    <g tabindex="0" filter="url(#drop-shadow)" transform="translate(300 0) scale(0.2 0.2) rotate(90)"
        transform-origin="center center"><svg xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 1000 1000" overflow="visible"
            shape-rendering="geometricPrecision" text-rendering="geometricPrecision">
            <style>
                text {
                    transform-origin: center center;
                }
            </style>
            <g transform="translate(0 0) scale(1.0) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="#d2dbed" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                    <path
                        d="M 0 500 L 200 500 L 250 300 L 350 700 L 450 300 L 550 700 L 650 300 L 750 700 L 800 500 L 1000 500"
                        fill="transparent" stroke="black" stroke-width="8"></path>
                </svg></g>
            <g transform="translate(-500 0) scale(0.2 0.2) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="#d2dbed" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                </svg></g>
            <g transform="translate(500 0) scale(0.2 0.2) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="blue" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                </svg></g>
        </svg></g>
    <g tabindex="0" filter="url(#drop-shadow)" transform="translate(0 0) scale(0.2 0.2) rotate(90)"
        transform-origin="center center"><svg xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 1000 1000" overflow="visible"
            shape-rendering="geometricPrecision" text-rendering="geometricPrecision">
            <style>
                text {
                    transform-origin: center center;
                }
            </style>
            <g transform="translate(0 0) scale(1.0) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="#d2dbed" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                    <path d="M 0 500 L 400 500 M 400 200 L 400 800 M 600 200 L 600 800 M 600 500 L 1000 500"
                        stroke="black" stroke-width="8"></path>
                </svg></g>
            <g transform="translate(-500 0) scale(0.2 0.2) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="#d2dbed" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                </svg></g>
            <g transform="translate(500 0) scale(0.2 0.2) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="blue" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                </svg></g>
        </svg></g>
    <g tabindex="0" filter="url(#drop-shadow)" transform="translate(-200 -300) scale(0.2 0.2) rotate(180)"
        transform-origin="center center"><svg xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 1000 1000" overflow="visible"
            shape-rendering="geometricPrecision" text-rendering="geometricPrecision">
            <style>
                text {
                    transform-origin: center center;
                }
            </style>
            <g transform="translate(0 0) scale(1.0) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="#d2dbed" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                    <g id="rings">
                        <circle cx="275" cy="500" r="75" fill="transparent" stroke="black" stroke-width="8"></circle>
                        <circle cx="425" cy="500" r="75" fill="transparent" stroke="black" stroke-width="8"></circle>
                        <circle cx="575" cy="500" r="75" fill="transparent" stroke="black" stroke-width="8"></circle>
                        <circle cx="725" cy="500" r="75" fill="transparent" stroke="black" stroke-width="8"></circle>
                    </g>

                    <rect x="100" y="500" width="800" height="100" fill="#d2dbed"></rect>

                    <path d="M 0 500 L 200 500" fill="transparent" stroke="black" stroke-width="8"></path>
                    <path d="M 800 500 L 1000 500" fill="transparent" stroke="black" stroke-width="8"></path>
                </svg></g>
            <g transform="translate(-500 0) scale(0.2 0.2) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="#d2dbed" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                </svg></g>
            <g transform="translate(500 0) scale(0.2 0.2) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="blue" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                </svg></g>
        </svg></g>
    <g tabindex="0" filter="url(#drop-shadow)" transform="translate(-400 0) scale(0.2 0.2) rotate(90)"
        transform-origin="center center"><svg xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 1000 1000" overflow="visible"
            shape-rendering="geometricPrecision" text-rendering="geometricPrecision">
            <style>
                text {
                    transform-origin: center center;
                }
            </style>
            <g transform="translate(0 0) scale(1.0) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="#d2dbed" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                    <text x="250" y="300" dominant-baseline="middle" text-anchor="middle" font-weight="200"
                        font-size="300">+</text>
                    <text x="750" y="300" dominant-baseline="middle" text-anchor="middle" font-weight="200"
                        font-size="300">-</text>
                    <path d="M 0 500 L 400 500 M 400 100 L 400 900 M 600 200 L 600 800 M 600 500 L 1000 500"
                        stroke="black" stroke-width="8"></path>
                </svg></g>
            <g transform="translate(-500 0) scale(0.2 0.2) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="#d2dbed" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                </svg></g>
            <g transform="translate(500 0) scale(0.2 0.2) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="blue" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                </svg></g>
        </svg></g>
    <g tabindex="0" filter="url(#drop-shadow)" transform="translate(0 500) scale(0.2 0.2) rotate(0)"
        transform-origin="center center"><svg xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 1000 1000" overflow="visible"
            shape-rendering="geometricPrecision" text-rendering="geometricPrecision">
            <style>
                text {
                    transform-origin: center center;
                }
            </style>
            <g transform="translate(0 0) scale(1.0) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="#d2dbed" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                    <path
                        d="M 500 0 L 500 300 M 100 300 L 900 300 M 200 475 L 800 475 M 300 650 L 700 650 M 400 825 L 600 825"
                        stroke="black" stroke-width="8"></path>
                </svg></g>
            <g transform="translate(0 -500) scale(0.2 0.2) rotate(0)" transform-origin="center center"><svg
                    xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                    viewBox="0 0 1000 1000" shape-rendering="geometricPrecision" text-rendering="geometricPrecision"
                    transform-origin="center center">
                    <rect rx="0" ry="0" width="1000" height="1000" fill="blue" stroke="blue" stroke-width="3"
                        vector-effect="non-scaling-stroke"></rect>
                </svg></g>
        </svg></g>
    <path d="M 100 400 L 100 200 L 200 200" stroke="blue" stroke-stroke="4" fill="transparent"
        vector-effect="non-scaling-stroke"></path>
    <path d="M 400 200 L 500 200 L 800 200 L 800 400" stroke="blue" stroke-stroke="4" fill="transparent"
        vector-effect="non-scaling-stroke"></path>
    <path d="M 800 400 L 800 200 L 500 200 L 500 400" stroke="blue" stroke-stroke="4" fill="transparent"
        vector-effect="non-scaling-stroke"></path>
    <path d="M 800 600 L 800 800 L 500 800 L 500 900" stroke="blue" stroke-stroke="4" fill="transparent"
        vector-effect="non-scaling-stroke"></path>
    <path d="M 500 900 L 500 800 L 500 600" stroke="blue" stroke-stroke="4" fill="transparent"
        vector-effect="non-scaling-stroke"></path>
    <path d="M 500 600 L 500 800 L 100 800 L 100 600" stroke="blue" stroke-stroke="4" fill="transparent"
        vector-effect="non-scaling-stroke"></path>
</svg>