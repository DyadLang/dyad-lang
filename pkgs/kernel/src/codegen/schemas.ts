import { Expression, ComponentDefinition } from "@juliacomputing/dyad-ast";
import { z } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import { instantiateModel, normalizeBaseType } from "../instantiate/index.js";
import {
  Problem,
  assertUnreachable,
  createError,
  Result,
  partialResult,
} from "@juliacomputing/dyad-common";
import { Either, Left, Right } from "purify-ts/Either";
import { ifEither } from "../result.js";
import { UnimplementedError } from "../workspace/errors.js";
import {
  getMaxAttribute,
  getMinAttribute,
  QueryHandler,
} from "../workspace/index.js";

const unexpectedType = createError("unexpected-type", "Unexpected type");

export function generateZod(
  model: ComponentDefinition,
  query: QueryHandler
): Result<z.ZodType<any, z.ZodTypeDef, any>> {
  return instantiateModel(model, {}, null, query).chain((instance) => {
    const shape: Record<string, z.ZodType<any, z.ZodTypeDef, any>> = {};
    const problems: Problem[] = [];

    for (const [cn, comp] of Object.entries(instance.parameters)) {
      switch (comp.type.resolves) {
        case "scalar":
          const base = normalizeBaseType(comp.type);
          const mods = comp.attributes;
          switch (base) {
            case "Real": {
              let schema = z.number();
              if (comp.doc_string !== null) {
                schema = schema.describe(comp.doc_string.value);
              }

              getMinAttribute(mods).ifJust((v) => {
                schema = schema.gte(v);
              });
              getMaxAttribute(mods).ifJust((v) => {
                schema = schema.lte(v);
              });
              shape[cn] = schema;

              comp.default.ifJust((x) => {
                ifEither(getRealEither(x), problems, (v) => {
                  shape[cn] = schema.default(v);
                });
              });
              break;
            }
            case "Integer": {
              let schema = z.number();

              getMinAttribute(mods).ifJust((v) => {
                schema = schema.gte(v);
              });
              getMaxAttribute(mods).ifJust((v) => {
                schema = schema.lte(v);
              });

              shape[cn] = schema;

              comp.default.ifJust((x) => {
                ifEither(getIntEither(x), problems, (v) => {
                  shape[cn] = schema.default(v);
                });
              });
              break;
            }
            case "String": {
              let schema = z.string();
              if (comp.doc_string !== null) {
                schema = schema.describe(comp.doc_string.value);
              }
              shape[cn] = schema;
              comp.default.ifJust((x) => {
                ifEither(getStringEither(x), problems, (v) => {
                  shape[cn] = schema.default(v);
                });
              });
              break;
            }
            case "Boolean": {
              let schema = z.boolean();
              if (comp.doc_string !== null) {
                schema = schema.describe(comp.doc_string.value);
              }
              shape[cn] = schema;
              comp.default.ifJust((x) => {
                ifEither(getBooleanEither(x), problems, (v) => {
                  shape[cn] = schema.default(v);
                });
              });
              break;
            }
            /* istanbul ignore next */
            default:
              assertUnreachable(base);
          }
          break;
        default:
          throw new UnimplementedError(
            "generateZod",
            `The generateZod function currently cannot handle ${comp.type.resolves} types`
          );
      }
    }
    return partialResult(z.object(shape), ...problems);
  });
}

export function generateSchema(
  model: ComponentDefinition,
  query: QueryHandler
) {
  return generateZod(model, query).map((v) => zodToJsonSchema(v));
}

export function getRealEither(e: Expression): Either<Problem, number> {
  if (e.type === "ilit" || e.type === "rlit") {
    return Right(e.value);
  } else {
    return Left(
      new unexpectedType(
        e.type,
        `Expected real value, but got literal of type ${e.type}`
      )
    );
  }
}

export function getIntEither(e: Expression): Either<Problem, number> {
  if (e.type === "ilit") {
    return Right(e.value);
  } else {
    return Left(
      new unexpectedType(
        e.type,
        `Expected integer value, but got literal of type ${e.type}`
      )
    );
  }
}

export function getBooleanEither(e: Expression): Either<Problem, boolean> {
  if (e.type === "blit") {
    return Right(e.value);
  } else {
    return Left(
      new unexpectedType(
        e.type,
        `Expected boolean value, but got literal of type ${e.type}`
      )
    );
  }
}

export function getStringEither(e: Expression): Either<Problem, string> {
  if (e.type === "slit") {
    return Right(e.value);
  } else {
    return Left(
      new unexpectedType(
        e.type,
        `Expected string value, but got literal of type ${e.type}`
      )
    );
  }
}
