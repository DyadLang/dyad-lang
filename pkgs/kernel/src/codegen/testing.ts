import { sourceKey } from "@juliacomputing/dyad-ast";
import {
  InMemoryLibraryProvider,
  stringifyProject,
} from "../providers/index.js";
import { Workspace } from "../workspace/workspace.js";
import { sourceFolder } from "@juliacomputing/dyad-common";
import { BlobReader, BlobWriter, ZipWriter } from "@zip.js/zip.js";
import { ZipProvider } from "../providers/zip.js";
import { CodeGenKey, ModuleName } from "./mtk/events.js";
import { FSHandler } from "./mtk/fshandler.js";
import { InMemoryFileSystem } from "../providers/memfs.js";
import { queryRawFiles } from "../workspace/index.js";

export const projectRLC = `
name="RLC"
uuid="abc123"
version="0.1.0"
authors=["Michael M. Tiller <michael.tiller@juliahub.com>"]`;

export async function loadWorkspaceToZip(
  libname: string,
  sources: Record<string, string>
) {
  const blobWriter = new BlobWriter();
  const writer = new ZipWriter(blobWriter);
  const addFile = async (filename: string, contents: string) => {
    const blob = new Blob([contents]);
    const reader = new BlobReader(blob);
    await writer.add(filename, reader);
  };
  await addFile(
    "./Project.toml",
    stringifyProject({
      name: libname,
      authors: [],
      uuid: "abc123-" + libname,
      version: "0.1.0",
    })
  );
  for (const [filename, contents] of Object.entries(sources)) {
    await addFile(`./${sourceFolder}/${filename}`, contents);
  }

  const workspace = await Workspace.create();
  await writer.close();

  const fileBlob = await blobWriter.getData();
  await fileBlob.arrayBuffer();

  const zip = new ZipProvider(fileBlob);

  const id = await workspace.registerProvider(zip);
  await workspace.waitForId(id);

  const files = workspace.query(queryRawFiles());
  const issues = files.flatMap((file) => file.problems);

  return { workspace, zip, issues };
}
/**
 * This function can be used for testing by easily provisioning a workspace and
 * loading it with some source files.
 * @param libname Library to add
 * @param sources Sources to add
 * @returns both the workspace and in memory library provider
 */
export async function loadWorkspace(
  libname: string,
  sources: Record<string, string>
) {
  const workspace = await Workspace.create();
  const inmem = await addLibrary(libname, workspace, sources);

  const files = workspace.query(queryRawFiles());
  const issues = files.flatMap((file) => file.problems);
  return { workspace, inmem, issues };
}

export async function addLibrary(
  libname: string,
  workspace: Workspace,
  sources: Record<string, string>
) {
  const inmem = new InMemoryLibraryProvider({
    name: libname,
    authors: [],
    uuid: "abc123-" + libname,
    version: "0.1.0",
  });
  const id = await workspace.registerProvider(inmem);
  await workspace.waitForId(id);

  for (const [file, source] of Object.entries(sources)) {
    const txn = await inmem.set(sourceKey(file, []), source);
    await workspace.waitForId(txn.transactionId);
  }
  return inmem;
}

export class SnapshotHandler extends FSHandler {
  constructor(protected expectedModules: Array<ModuleName>) {
    super(new InMemoryFileSystem());
  }
  async modules(all: Array<ModuleName>): Promise<void> {
    await super.modules(all);
    expect(all).toEqual(this.expectedModules);
  }
  async source(key: CodeGenKey, code: string): Promise<void> {
    await super.source(key, code);
    expect({ key, code }).toMatchSnapshot();
  }
  async close(): Promise<void> {
    expect(this.fs instanceof InMemoryFileSystem).toEqual(true);
    if (this.fs instanceof InMemoryFileSystem) {
      const fs = this.fs;
      expect(fs.toJSON()).toMatchSnapshot();
    }
  }
}
