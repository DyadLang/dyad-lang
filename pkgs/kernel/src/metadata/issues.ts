import { ProblemSpan, TextProblem } from "@juliacomputing/dyad-ast";
import { z } from "zod";

/* istanbul ignore next @preserve */
export function zodIssue2Problem(context: string, schema: string) {
  return (issue: z.ZodIssue, span: ProblemSpan): TextProblem => {
    let title = "Zod Issue";
    let instance = "";
    let extra: any = {};
    switch (issue.code) {
      case "custom":
        title = "Custom Zod Type";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = issue.params;
        break;
      case "invalid_arguments":
        title = "Invalid Arguments";
        instance = issue.path.map((x) => x.toString()).join(".");
        break;
      case "invalid_date":
        title = "Invalid Date";
        instance = issue.path.map((x) => x.toString()).join(".");
        break;
      case "invalid_enum_value":
        title = "Invalid Enum Value";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = { options: issue.options };
        break;
      case "invalid_intersection_types":
        title = "Invalid Intersection Types";
        instance = issue.path.map((x) => x.toString()).join(".");
        break;
      case "invalid_literal":
        title = "Invalid Literal";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = { expected: issue.expected, received: issue.received };
        break;
      case "invalid_return_type":
        title = "Invalid Return Type";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = { returnTypeError: issue.returnTypeError };
        break;
      case "invalid_string":
        title = "Invalid String";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = { validation: issue.validation };
        break;
      case "invalid_type":
        title = "Invalid Type";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = { expected: issue.expected, received: issue.received };
        break;
      case "invalid_union":
        title = "Invalid Union";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = { unionErrors: issue.unionErrors };
        break;
      case "invalid_union_discriminator":
        title = "Invalid Union Discriminator";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = { options: issue.options };
        break;
      case "not_finite":
        title = "Not Finite";
        instance = issue.path.map((x) => x.toString()).join(".");
        break;
      case "not_multiple_of":
        title = "Not Multiple Of";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = { multipleOf: issue.multipleOf };
        break;
      case "too_big":
        title = "Too Big";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = {
          maximum: issue.maximum,
          inclusive: issue.inclusive,
          exact: issue.exact,
          type: issue.type,
        };
        break;
      case "too_small":
        title = "Too Small";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = {
          minimum: issue.minimum,
          inclusive: issue.inclusive,
          exact: issue.exact,
          type: issue.type,
        };
        break;
      case "unrecognized_keys":
        title = "Unrecognized Keys";
        instance = issue.path.map((x) => x.toString()).join(".");
        extra = { keys: issue.keys };
        break;
      default:
        title = "Unrecognized Zod Issue";
        instance = "unrecognized";
        break;
    }
    let details = `In ${context} while validating the ${instance} field against ${schema} schema, got: ${
      issue.message
    }: ${JSON.stringify(extra)}`;

    return {
      severity: issue.fatal ? "error" : "warning",
      type: `zod:${issue.code}`,
      title,
      // The instance is the context (calling function), schema validated and then generally the path
      instance: `${context}/${schema}::${instance}`,
      details,
      extra: { ...extra, ...span },
    };
  };
}
