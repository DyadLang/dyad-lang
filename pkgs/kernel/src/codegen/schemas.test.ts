import { rlc_sources } from "./samples/rlc.js";
import { queryType } from "../workspace/index.js";
import { isComponentDefinition } from "@juliacomputing/dyad-ast";
import { loadWorkspace, loadWorkspaceToZip } from "./testing.js";
import { generateSchema, generateZod } from "./schemas.js";
import { zodIssue2Problem } from "../metadata/issues.js";
import { SafeParseReturnType } from "zod";
import { assertHasResult, assertJust } from "../workspace/utils.test.js";

const schema_sources = {
  "complex.dyad": `type Pressure = Real
type Resistance = Real(min=0)
component Complex
  parameter v::Real(min=0)
  parameter p::Pressure = 101326
  parameter sw::Boolean
  parameter dataFile::String
  # Number of elements to create (this should end up in the schema description)
  parameter n::Integer(min=1,max=5)
  parameter R::Resistance
end`,
};

describe("Test Schema generation", () => {
  test("Test schema generation for resistor model", async () => {
    /* Create a workspace with a single library */
    const { workspace } = await loadWorkspace("RLC", rlc_sources);

    const def = workspace.query(queryType("RLC", [], "Resistor"));
    expect(def.isJust()).toEqual(true);
    const definition = def.unsafeCoerce();
    expect(isComponentDefinition(definition)).toEqual(true);
    if (isComponentDefinition(definition)) {
      const zod = assertHasResult(
        generateZod(definition, workspace.query.bind(workspace))
      );
      const v1 = zod.safeParse({ R: 10 });
      expect(v1.success).toEqual(true);
      const v2 = zod.safeParse({});
      expect(v2.success).toEqual(false);
      const v3 = zod.safeParse({ R: 10, T: 20 });
      expect(v3.success).toEqual(true);

      const schema = assertHasResult(
        generateSchema(definition, workspace.query.bind(workspace))
      );
      expect(schema).toMatchSnapshot();
    }
  });
  test("Test schema generation for complex model", async () => {
    /* Create a workspace with a single library */
    const { workspace, issues } = await loadWorkspaceToZip(
      "Schemas",
      schema_sources
    );

    expect(issues).toEqual([]);
    const def = assertJust(
      workspace.query(queryType("Schemas", [], "Complex"))
    );
    expect(isComponentDefinition(def)).toEqual(true);
    if (isComponentDefinition(def)) {
      const v1IssueMap = zodIssue2Problem("v1", "generated");
      const base = {
        v: 0,
        p: 101325,
        sw: true,
        dataFile: "foo.txt",
        n: 3,
        R: 100,
      };
      const checkIssues = (v: SafeParseReturnType<any, any>) => {
        if (!v.success) {
          expect(
            v.error.issues.map((x) => v1IssueMap(x, { file: null, span: null }))
          ).toEqual([]);
        }
        expect(v.success).toEqual(true);
      };
      const zod = assertHasResult(
        generateZod(def, workspace.query.bind(workspace))
      );
      const v1 = zod.safeParse(base);
      checkIssues(v1);
      const v2 = zod.safeParse({});
      expect(v2.success).toEqual(false);
      const v3 = zod.safeParse({ ...base, v: -1 });
      expect(v3.success).toEqual(false);
      const v4 = zod.safeParse({ ...base, n: 6 });
      expect(v4.success).toEqual(false);
      const v5 = zod.safeParse({ ...base, n: 6.5 });
      expect(v5.success).toEqual(false);
      const v6 = zod.safeParse({ ...base, R: -100 });
      expect(v6.success).toEqual(false);

      const schema = assertHasResult(
        generateSchema(def, workspace.query.bind(workspace))
      );
      expect(schema).toMatchSnapshot();
    }
  });
});
