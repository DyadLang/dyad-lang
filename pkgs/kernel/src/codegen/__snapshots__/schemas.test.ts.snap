// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Schema generation Test schema generation for complex model 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "R": {
      "minimum": 0,
      "type": "number",
    },
    "dataFile": {
      "type": "string",
    },
    "n": {
      "maximum": 5,
      "minimum": 1,
      "type": "number",
    },
    "p": {
      "default": 101326,
      "type": "number",
    },
    "sw": {
      "type": "boolean",
    },
    "v": {
      "minimum": 0,
      "type": "number",
    },
  },
  "required": [
    "v",
    "sw",
    "dataFile",
    "n",
    "R",
  ],
  "type": "object",
}
`;

exports[`Test Schema generation Test schema generation for resistor model 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "R": {
      "description": "Resistance of this Resistor",
      "minimum": 0,
      "type": "number",
    },
  },
  "required": [
    "R",
  ],
  "type": "object",
}
`;
